homeassistant/components/withings/common.py

```diff
@@ -89,184 +370,278 @@ def __init__(
+            await asyncio.sleep(self._notify_subscribe_delay.total_seconds())
+            await self._async_call_retry(
+                lambda a=appli: self._api.notify_subscribe(self._webhook_config.url, a)
+            )
 
-            # Oauth2 config flow failed to authenticate.
-            if NOT_AUTHENTICATED_ERROR.match(str(ex)):
-                raise NotAuthenticatedError(ex)
+    async def async_unsubscribe_webhook(self) -> None:
+        """Unsubscribe webhook from withings data updates."""
+        # Get the current webhooks.
+        response = await self._async_call_retry(self._api.notify_list)
 
-            # Probably a network error.
-            WithingsDataManager.print_service_unavailable()
-            raise PlatformNotReady(ex)
+        # Revoke subscriptions.
+        for profile in response.profiles:
+            _LOGGER.info(
+                "Unsubscribing %s for %s.", profile.callbackurl, profile.appli,
+            )
+            await self._async_call_retry(
+                lambda p=profile: self._api.notify_revoke(p.callbackurl, p.appli)
+            )
 
-    async def check_authenticated(self) -> bool:
+    async def async_check_authenticated(self) -> bool:
         """Check if the user is authenticated."""
+        return bool(await self._async_call_retry(self._api.user_get_device))
 
-        def function():
-            return bool(self._api.user_get_device())
-
-        return await self.call(function)
+    async def async_update_all(self) -> None:
+        """Update all withings data."""
+        _LOGGER.info("Updating all withings data.")
+        await self.async_update_measure()
+        await self.async_update_sleep_summary()
 
-    async def update_measures(self) -> MeasureGetMeasResponse:
+    async def async_update_measure(self) -> None:
         """Update the measures data."""
+        _LOGGER.info("Updating withings measures.")
 
-        def function():
-            return self._api.measure_get_meas()
-
-        self._measures = await self.call(function, throttle_domain="update_measures")
-
-        return self._measures
-
-    async def update_sleep(self) -> SleepGetResponse:
-        """Update the sleep data."""
-        end_date = dt.now()
-        start_date = end_date - datetime.timedelta(hours=2)
+        response = await self._async_call_retry(self._api.measure_get_meas)
 
-        def function():
-            return self._api.sleep_get(startdate=start_date, enddate=end_date)
-
-        self._sleep = await self.call(function, throttle_domain="update_sleep")
+        groups = query_measure_groups(
+            response, MeasureTypes.ANY, MeasureGroupAttribs.UNAMBIGUOUS
+        )
 
-        return self._sleep
+        for group in groups:
+            for measure in group.measures:
+                value = float(measure.value * pow(10, measure.unit))
+                await self._async_set_data_for_measure_type(
+                    measure.type, round(value, 2)
+                )
```

Comment:
We should avoid doing `await` inside a for-loop. Can we parallelize this and do it in a gather?

```python
await asyncio.gather(*[
    self._async_set_data_for_measure_type(
        measure.type, round(float(measure.value * pow(10, measure.unit)), 2))
    for group in groups
    for measure in group.measures
])
```